bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/RawMirrorRequest.java

```diff
@@ -46,39 +46,43 @@ public RawMirrorRequest(IArtifactDescriptor sourceDescriptor, IArtifactDescripto
 
 	@Override
 	public void perform(IArtifactRepository sourceRepository, IProgressMonitor monitor) {
-		monitor.subTask(NLS.bind(Messages.downloading, getArtifactKey().getId()));
-		setSourceRepository(sourceRepository);
-		// Do we already have the descriptor in the target?
-		if (target.contains(targetDescriptor)) {
-			setResult(new Status(IStatus.INFO, Activator.ID, NLS.bind(Messages.mirror_alreadyExists, targetDescriptor, target)));
-			return;
-		}
-		// Does the source actually have the descriptor?
-		if (!source.contains(getArtifactDescriptor())) {
-			setResult(new Status(IStatus.ERROR, Activator.ID, NLS.bind(Messages.artifact_not_found, getArtifactKey())));
-			return;
-		}
-		IStatus status = transfer(targetDescriptor, sourceDescriptor, monitor);
+		SubMonitor subMon = SubMonitor.convert(monitor,NLS.bind(Messages.downloading, getArtifactKey().getId()),1);
```

Comment:
Converting the monitor into a `SubMonitor` with work one does not look necessary here and it can just be passed directly.
Furthermore I think it is not necessary to add the try-catch-block to ensure done is called in any ways.
Just calling done below is sufficient and if the method throws the method returns any way and a caller that catches an exception either calls done on the passed monitor or (should) call split which also cleans up children.

Reply:
> Furthermore I think it is not necessary to add the try-catch-block to ensure done is called in any ways

Correct, IProgressmonitor states:

CALLER/CALLEE RESPONSIBILITIES:

Methods that receive an [IProgressMonitor](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82IProgressMonitor) ("callees") must either obey the following conventions or include JavaDoc explaining how it differs from these rules. The called method:

Will call [beginTask](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82beginTask) on the argument 0 or 1 times, at its option.
Will not promise to invoke [done()](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82done%E2%98%82) on the monitor.
Will not call [setCanceled](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82setCanceled) on the monitor.
May rely on the monitor not being null.
May rely on the monitor ignoring the string passed to [beginTask](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82beginTask).

Reply:
> Furthermore I think it is not necessary to add the try-catch-block to ensure done is called in any ways

Correct, IProgressmonitor states:

CALLER/CALLEE RESPONSIBILITIES:

Methods that receive an [IProgressMonitor](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82IProgressMonitor) ("callees") must either obey the following conventions or include JavaDoc explaining how it differs from these rules. The called method:

Will call [beginTask](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82beginTask) on the argument 0 or 1 times, at its option.
Will not promise to invoke [done()](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82done%E2%98%82) on the monitor.
Will not call [setCanceled](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82setCanceled) on the monitor.
May rely on the monitor not being null.
May rely on the monitor ignoring the string passed to [beginTask](eclipse-javadoc:%E2%98%82=org.eclipse.equinox.common/src%3Corg.eclipse.core.runtime%7BIProgressMonitor.java%E2%98%83IProgressMonitor%E2%98%82%E2%98%82beginTask).

Reply:
> furthermore I think it is not necessary to add the try-catch-block to ensure done is called in any ways.

Yep, see javadoc of IProgressMonitor ...

Reply:
By passing a submonitor with progress one and unconditional calling done() the progress is unconditional increased which ensures to use up all the reserved progress. That is technically not necessary but ensures a user readable progress from 0 to 100% . This commit fixes the error and proofs that with junit test. This pattern is easy to understand and not error prone. I am not interested in further "optimizations" that "may" work under special conditions.

Reply:
I just wanted to note that when one splits a monitor previous aquired "splits" will be completed automatically, so if the toplevel is calling `done()` and the other code is always using `split()` progress si still reported correctly in every case.

According to `IProgressMonitor` javadoc the called method: 

1. Will call beginTask on the argument 0 or 1 times, at its option.
2. Will not promise to invoke done() on the monitor.
3. Will not call setCanceled on the monitor.
4. May rely on the monitor not being null.
5. May rely on the monitor ignoring the string passed to beginTask.

so calling done is of course _allowed_, but not _required_ by a method receiving a monitor.